plugin-viewport/src/main/kotlin/com/mapbox/maps/plugin/viewport/experimental/ViewportPluginImpl.kt

```diff
@@ -0,0 +1,210 @@
+    if (status.getTargetState() === targetState) {
+      completionListener?.onComplete(false)
+      return
+    }
+    val fromState = status.getTargetState()
+    currentCancelable?.cancel()
+    currentCancelable = null
+    if (!states.contains(targetState)) {
+      throw IllegalStateException("Target ViewportState is not added.")
+    }
+
+    // get the transition (or default) for the from and to state
+    val transition = getTransition(fromState, targetState) ?: defaultTransition
+
+    // run the transition
+    var completionBlockInvoked = false
+    val transitionCancelable = transition.run(fromState, targetState) { isCanceled ->
+      completionBlockInvoked = true
+      if (!isCanceled) {
+        // transfer camera updating responsibility to targetState
+        currentCancelable = targetState.startUpdatingCamera()
+        updateStatus(
+          ViewportStatus.State(targetState),
+          VIEWPORT_STATUS_OBSERVER_REASON_PROGRAMMATIC
+        )
+      }
+      completionListener?.onComplete(isCanceled)
+    }
+    // since it's possible that a transition might invoke its
+    // completion block synchronously, we'll only store the
+    // transition cancelable if the transition is not complete
+    // so that we don't clobber the targetState cancelable.
+    if (!completionBlockInvoked) {
+      currentCancelable = transitionCancelable
+      updateStatus(
+        ViewportStatus.Transition(transition, fromState, targetState),
+        VIEWPORT_STATUS_OBSERVER_REASON_PROGRAMMATIC
+      )
+    }
+  }
+
+  override fun toIdle() {
+    if (status.getTargetState() == null) return
+    currentCancelable?.cancel()
+    currentCancelable = null
+    updateStatus(ViewportStatus.State(null), VIEWPORT_STATUS_OBSERVER_REASON_PROGRAMMATIC)
+  }
+
+  private fun notifyStatusChanged(
+    previousStatus: ViewportStatus?,
+    currentStatus: ViewportStatus?,
+    reason: String
+  ) {
+    registeredStatusObservers.forEach {
+      it.onViewportStateChanged(
+        previousStatus,
+        currentStatus,
+        reason
+      )
+    }
+  }
+
+  override fun addState(state: ViewportState) {
+    internalStates.add(state)
+  }
+
+  override fun removeState(state: ViewportState) {
+    if (status.getTargetState() === state) {
+      toIdle()
+    }
+    internalStates.remove(state)
+  }
+
+  // this transition is used unless overridden by one of the registered transitions
+  override lateinit var defaultTransition: ViewportTransition
+
+  override fun setTransition(
+    transition: ViewportTransition,
+    from: ViewportState?,
+    to: ViewportState
+  ) {
+    transitions[Pair(from, to)] = transition
+  }
+
+  override fun getTransition(from: ViewportState?, to: ViewportState): ViewportTransition? {
+    return transitions[Pair(from, to)]
+  }
+
+  override fun removeTransition(from: ViewportState?, to: ViewportState) {
+    transitions.remove(Pair(from, to))
+  }
+
+  override var transitionsToIdleUponUserInteraction: Boolean = true
```

Comment:
How about introducing a `ViewportPluginOptions` for this and future customizations of the plugin itself, next to customizing respective states?

Alternatively, this should be moved to the state layer and each state should handle its own gesture interactions, that might better fit the proposed pattern.

Reply:
Interesting idea, the transitionsToIdleUponUserInteraction is just a convenient method that allow user to disrupt the state or transition with gestures. More complex logic shall be implemented by the user to handle customised gesture inputs.

I would vote for adding these properties as a separate options class for ViewportPlugin, wdyt? cc @macdrevx